### PR TITLE
-Issue: #205: Merging (Identically Specified) MinHashLSH objects

### PR DESCRIPTION
**Pull request by rupeshkumaar**
Merging two MinHashLSH of the same length (num_perm).

### CODE REVIEW

**Comment by rupeshkumaar** [comment]
@ekzhu I have raised a PR. I have tried to cover all the points that we had discussed. Please feel free to engage and let me know if any modification is required. Thank you!

File: datasketch/lsh.py [hunk 1]
```diff
@@ -282,6 +301,36 @@ def _insert(
         for H, hashtable in zip(Hs, self.hashtables):
             hashtable.insert(H, key, buffer=buffer)
 
+    def __eq__(self, other:MinHashLSH) -> bool:
```
Comment (ekzhu) [inline datasketch/lsh.py:304]: Would be better to call it a different name. `__eq__` implies complete equivalence. This is not the case as it only checks for initialization parameters.

File: datasketch/lsh.py [hunk 2]
```diff
@@ -282,6 +301,36 @@ def _insert(
...
+        """
+        Returns:
+            bool: If the two MinHashLSH has equal num_perm then two are equivalent.
+        """
+        return (
```
Comment (ekzhu) [inline datasketch/lsh.py:309]: More checks need to be made. E.g., the b and r parameters.

File: test/test_lsh.py [hunk 3]
```diff
@@ -240,6 +240,45 @@ def test_get_counts(self):
         for table in counts:
             self.assertEqual(sum(table.values()), 2)
 
+    def test_merge(self):
```
Comment (ekzhu) [inline test/test_lsh.py:243]: Can you add test for Redis storage backend too?

File: datasketch/lsh.py (hunk 1 above)
Comment (rupeshkumaar) [reply datasketch/lsh.py:304]: Alright, will change it to something else.

File: datasketch/lsh.py (hunk 2 above)
Comment (rupeshkumaar) [reply datasketch/lsh.py:309]: Added, b and r parameters. Please let me know if you think we need any other parameters too.

File: test/test_lsh.py (hunk 3 above)
Comment (rupeshkumaar) [reply test/test_lsh.py:243]: Okay, will do it.

**Comment by rupeshkumaar** [comment]
@ekzhu Resolved the conversations. Can you please go through the changes. Thank you. Let me know if there are any other changes.

File: datasketch/lsh.py [hunk 4]
```diff
@@ -282,6 +301,38 @@ def _insert(
...
+        """
+        Returns:
+            bool: If the two MinHashLSH has equal num_perm, band size and size of each bands then two are equivalent.
+        """
+        return (
+            type(self) is type(other) and
+            self.h == other.h and
+            self.b == other.b and
+            self.r == other.r
+        )
```
Comment (ekzhu) [inline datasketch/lsh.py:314]: There are also:
1. the hash function
2. the weights

Comment (rupeshkumaar) [reply datasketch/lsh.py:314]: alright! added it as well.

Comment (ekzhu) [reply datasketch/lsh.py:315]: Hmm consider this for function equivalence?
https://stackoverflow.com/questions/20059011/check-if-two-python-functions-are-equal


 Probably a good idea to add some tests for some error cases.

File: test/test_lsh.py [hunk 5]
```diff
@@ -240,6 +240,97 @@ def test_get_counts(self):
...
+                items.extend(t[H])
+            self.assertTrue("c" in items)
+            self.assertTrue("d" in items)
+        self.assertTrue("a" in lsh1)
+        self.assertTrue("b" in lsh1)
+        self.assertTrue("c" in lsh1)
+        self.assertTrue("d" in lsh1)
+        for i, H in enumerate(lsh1.keys["c"]):
+            self.assertTrue("c" in lsh1.hashtables[i][H])
+
```
Comment (ekzhu) [inline test/test_lsh.py:274]: Can you also add tests for merging two LSH when duplicate keys exist, but check_duplicate is False. Do this for all storage layer.

File: datasketch/lsh.py [hunk 6]
```diff
@@ -282,6 +301,39 @@ def _insert(
...
+            self.h == other.h and
+            self.b == other.b and
+            self.r == other.r and
+            type(self.hashfunc) == type(other.hashfunc)
+        )
+
+    def _merge(
+        self,
+        other: MinHashLSH,
+        check_disjointness: bool = False,
```
Comment (ekzhu) [inline datasketch/lsh.py:320]:
```suggestion
        check_disjoint: bool = False,
```

File: datasketch/lsh.py [hunk 7]
```diff
@@ -226,6 +226,25 @@ def insert(
...
+            self,
+            other: MinHashLSH,
+            check_disjointness: bool = False      
+    ):
+        """Merge the other MinHashLSH with this one, making this one the union
+        of both the MinHashLSH.
+
+        Args:
+            other (MinHashLSH): The other MinHashLSH.
+            check_duplication (bool): To avoid duplicate keys in the storage
```
Comment (ekzhu) [inline datasketch/lsh.py:239]:
```suggestion
            check_duplicate_keys (bool): Check if the is any overlapping keys before merging and raise if there are any. 
```

File: datasketch/lsh.py [hunk 8]
```diff
@@ -226,6 +226,25 @@ def insert(
...
+                (`default=True`)
+
+        Raises:
+            ValueError: If the two MinHashLSH have different initialization
```
Comment (ekzhu) [inline datasketch/lsh.py:243]:
```suggestion
            ValueError: If the two MinHashLSH have different initialization parameters, or if `check_overlap` is `True` and there are overlapping keys.
```

**Comment by junrae6454** [comment]
@ekzhu When will this commit be merged?

**Comment by ekzhu** [comment]
@junrae6454 Still waiting for @rupeshkumaar's changes. Unless you want to help finishing this?

File: test/test_lsh.py (hunk 5 above)
Comment (rupeshkumaar) [reply test/test_lsh.py:274]: The default one is the False. But I am adding a test case specifying the check_disjointness as False. BTW the argument name is check_disjointness. Earlier, it was check_duplicate but we discussed it to change to check_disjointness.

File: datasketch/lsh.py (hunk 6 above)
Comment (rupeshkumaar) [reply datasketch/lsh.py:320]: The argument is changed to check_overlap

File: datasketch/lsh.py (hunk 7 above)
Comment (rupeshkumaar) [reply datasketch/lsh.py:239]: This has also changed. Thanks @ekzhu

**Comment by rupeshkumaar** [comment]
> Could you also add "3.12" to the tested python version:
> 
> https://github.com/ekzhu/datasketch/blob/master/.github/workflows/test.yml#L11

This is also done.

**Comment by ekzhu** [comment]
> > Could you also add "3.12" to the tested python version:
> > https://github.com/ekzhu/datasketch/blob/master/.github/workflows/test.yml#L11
> 
> This is also done.

Thanks. Let me know when the changes are pushed

File: datasketch/lsh.py (hunk 4 above)
Comment (rupeshkumaar) [reply datasketch/lsh.py:315]: I have checked out the link for function equivalence. It was too vague to follow. And I am not sure if we can use the inspect to use the source code of the function to compare. I have tried using two different hash functions one was the in-buit hash method and other md5 hash and sha1. And it worked properly for the normal storage option, but when I tested it for redis I was getting an error since redis key is in bytes and the hash function we are passing. Sharing the snippet. 

`from datasketch.hashfunc import sha1_hash32
lsh3 = MinHashLSH(threshold=0.5, num_perm=128, storage_config={
    'type': 'redis', 'redis': {'host': 'localhost', 'port': 6379}
}, hashfunc=sha1_hash32)`
while trying to insert the minhash object I was getting the TypeError saying **can't concat _hashlib.HASH to bytes** 
Can you please help out here? @ekzhu

**Comment by rupeshkumaar** [comment]
Please do let me know, @ekzhu about the hashfunc for redis that I have asked. Thank you. Logging out for now. Once this is done I will push the changes.

File: datasketch/lsh.py (hunk 4 above)
Comment (ekzhu) [reply datasketch/lsh.py:315]: I think what you have done is great already. If we don't have a precise definition of hash function equivalence, we just don't add the check to make the API well-defined.

Comment (ekzhu) [reply datasketch/lsh.py:315]: Regarding the issue with hashfunc, the hash function should input bytes and output bytes. The built-in `sha1_hash32` inputs bytes and output integer -- it is used for `MinHash`, not for `MinHashLSH`.

Comment (rupeshkumaar) [reply datasketch/lsh.py:315]: We can remove it and I can do a bit of cleaning and can push the changes but can you help me understanding since we would be facing this TypeError regardless. So, if we are trying to insert any MinHash object into a MinHashLSH object with a hash function defined and the storage options is redis we should be facing it. Or, is there anything that I am doing wrong? I will push the changes today. Thanks @ekzhu

Comment (ekzhu) [reply datasketch/lsh.py:315]: BTW, sorry for the delayed response...

Comment (rupeshkumaar) [reply datasketch/lsh.py:315]: No issues at all @ekzhu. Just to understand it we can only use the hash functions that either outputs the hash in bytes or a custom one to do the same. Please correct me if I am wrong.

Comment (ekzhu) [reply datasketch/lsh.py:315]: You should not have `TypeError`. The default hash function for `MinHashLSH` should work. `sha1_hash32` is for `MinHash` not for `MinHashLSH`. 

`MinHash`'s hash function is not used in `MinHashLSH`.

Comment (rupeshkumaar) [reply datasketch/lsh.py:315]: Okay, got it. I will push the changes once done today.

**Comment by rupeshkumaar** [comment]
@ekzhu Please review.

File: datasketch/lsh.py [hunk 9]
```diff
@@ -226,6 +226,25 @@ def insert(
...
+            other: MinHashLSH,
+            check_overlap: bool = False      
+    ):
+        """Merge the other MinHashLSH with this one, making this one the union
+        of both the MinHashLSH.
+
+        Args:
+            other (MinHashLSH): The other MinHashLSH.
+            check_overlap (bool): Check if there are any overlapping keys before merging and raise if there are any.
+                (`default=True`)
```
Comment (ekzhu) [inline datasketch/lsh.py:240]:
```suggestion
                (`default=False`)
```

File: datasketch/lsh.py [hunk 10]
```diff
@@ -226,6 +226,25 @@ def insert(
...
         self._insert(key, minhash, check_duplication=check_duplication, buffer=False)
 
+    def merge(
+            self,
+            other: MinHashLSH,
+            check_overlap: bool = False      
+    ):
+        """Merge the other MinHashLSH with this one, making this one the union
+        of both the MinHashLSH.
+
```
Comment (ekzhu) [inline datasketch/lsh.py:236]: Need to mention what initialization parameters are checked, and what are not checked (e.g., `hash_func` is not checked).

File: datasketch/lsh.py [hunk 11]
```diff
@@ -226,6 +226,25 @@ def insert(
         """
         self._insert(key, minhash, check_duplication=check_duplication, buffer=False)
 
+    def merge(
+            self,
+            other: MinHashLSH,
+            check_overlap: bool = False      
+    ):
+        """Merge the other MinHashLSH with this one, making this one the union
+        of both the MinHashLSH.
```
Comment (ekzhu) [inline datasketch/lsh.py:235]:
```suggestion
        of both.
```

File: docs/lsh.rst [hunk 12]
```diff
@@ -77,6 +77,14 @@ plotting code.
 .. figure:: /_static/lsh_benchmark.png
    :alt: MinHashLSH Benchmark
 
+You can merge two MinHashLSH object using the ``merge`` function. This
```
Comment (ekzhu) [inline docs/lsh.rst:80]:
```suggestion
You can merge two MinHashLSH indexes to create a union index using the ``merge`` method. This
```

**Review by ekzhu — CHANGES_REQUESTED** [review]

**Comment by ekzhu** [comment]
Also, to fix the tests failure for now, remove 3.12 from python version.

**Comment by rupeshkumaar** [comment]
Please review @ekzhu

**Review by ekzhu — APPROVED** [review]

**Comment by ekzhu** [comment]
@rupeshkumaar merged, thank you for the hard work!!

**Comment by ekzhu** [comment]
@rupeshkumaar if you are interested, you can take a look at this issue #236 that broke the library for python 3.12
